test/http_header_end2end_test.go

```diff
@@ -0,0 +1,216 @@
+				"content-type", "application/grpc",
+			},
+			errCode: codes.PermissionDenied,
+		},
+		{
+			name: "malformed grpc-status",
+			header: []string{
+				":status", "502",
+				"content-type", "application/grpc",
+				"grpc-status", "abc",
+			},
+			errCode: codes.Internal,
+		},
+		{
+			name: "Malformed grpc-tags-bin field",
+			header: []string{
+				":status", "502",
+				"content-type", "application/grpc",
+				"grpc-status", "0",
+				"grpc-tags-bin", "???",
+			},
+			errCode: codes.Unavailable,
+		},
+		{
+			name: "gRPC status error",
+			header: []string{
+				":status", "502",
+				"content-type", "application/grpc",
+				"grpc-status", "3",
+			},
+			errCode: codes.Unavailable,
+		},
+	} {
+		t.Run(test.name, func(t *testing.T) {
+			if err := doHTTPHeaderTest(t, test.errCode, test.header); err != nil {
+				t.Error(err)
+			}
+		})
+	}
+}
+
+// Testing non-Trailers-only Trailers (delivered in second HEADERS frame)
+func (s) TestHTTPHeaderFrameErrorHandlingNormalTrailer(t *testing.T) {
+	tests := []struct {
+		name           string
+		responseHeader []string
+		trailer        []string
+		errCode        codes.Code
+	}{
+		{
+			name: "trailer missing grpc-status",
+			responseHeader: []string{
+				":status", "200",
+				"content-type", "application/grpc",
+			},
+			trailer: []string{
+				// trailer missing grpc-status
+				":status", "502",
+			},
+			errCode: codes.Unavailable,
+		},
+		{
+			name: "malformed grpc-status-details-bin field with status 404",
+			responseHeader: []string{
+				":status", "404",
+				"content-type", "application/grpc",
+			},
+			trailer: []string{
+				// malformed grpc-status-details-bin field
+				"grpc-status", "0",
+				"grpc-status-details-bin", "????",
+			},
+			errCode: codes.Unimplemented,
+		},
+		{
+			name: "malformed grpc-status-details-bin field with status 200",
+			responseHeader: []string{
+				":status", "200",
+				"content-type", "application/grpc",
+			},
+			trailer: []string{
+				// malformed grpc-status-details-bin field
+				"grpc-status", "0",
+				"grpc-status-details-bin", "????",
+			},
+			errCode: codes.Internal,
+		},
+	}
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			if err := doHTTPHeaderTest(t, test.errCode, test.responseHeader, test.trailer); err != nil {
+				t.Error(err)
+			}
+		})
+
+	}
+}
+
+func (s) TestHTTPHeaderFrameErrorHandlingMoreThanTwoHeaders(t *testing.T) {
+	header := []string{
+		":status", "200",
+		"content-type", "application/grpc",
+	}
+	if err := doHTTPHeaderTest(t, codes.Internal, header, header, header); err != nil {
+		t.Fatal(err)
+	}
+}
+
+func doHTTPHeaderTest(t *testing.T, errCode codes.Code, headerFields ...[]string) error {
+	lis, err := net.Listen("tcp", "localhost:0")
+	if err != nil {
+		return fmt.Errorf("listening on %q: %v", "localhost:0", err)
+	}
+	defer lis.Close()
+	server := &httpServer{
+		responses: []httpServerResponse{{trailers: headerFields}},
+	}
+	server.start(t, lis)
```

Comment:
Can we create a separate function for the setup steps and mark it with a `t.Helper()`:

```golang
func startServer(t *testing.T, headerFields ...[]string) (serverAddr string, cleanup func()) {
	t.Helper()
	
	lis, err := net.Listen("tcp", "localhost:0")
	if err != nil {
		return t.Fatalf("listening on %q: %v", "localhost:0", err)
	}
	server := &httpServer{responses: []httpServerResponse{{trailers: headerFields}}}
	server.start(t, lis)
	return lis.Addr().String(), func() { lis.Close()}
}
```

If we do the above, we wouldn't have to pass `testing.T` to `doHTTPHeaderTest()`, and it would contain only the actual test. We could have to pass the `serverAddr` returned from `startServer()` though.

Reply:
sounds good. that way we can use t.Helper() for all the server setup